### PR TITLE
feat: cherry-pick #2960 into AKS official branch

### PR DESCRIPTION
**Pull request by norshtein**
**Reason for Change**:

This PR cherry-picks #2960 into AKS official branch. The commit enables BYO identity for kubelet, so after this commit cherry-picked, kubelet's identity can be created by AKS RP via Go client and then provide this identity to AKS-Engine. AKS RP don't need to rely on AKS-Engine generated ARM template to create the identity. 

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

**Comment by norshtein** [comment]
@jackfrancis could you please help merge the PR? @yangl900 @xuto2 FYI.

File: pkg/api/types.go [hunk 1]
```diff
@@ -1970,25 +1970,8 @@ func (k *KubernetesConfig) SystemAssignedIDEnabled() bool {
...
-}
-
-// GetUserAssignedClientID returns the user assigned client ID if it is enabled.
-func (k *KubernetesConfig) GetUserAssignedClientID() string {
-	if k.UserAssignedClientIDEnabled() {
-		return k.UserAssignedClientID
-	}
-	return ""
+func (k *KubernetesConfig) ShouldCreateNewUserAssignedIdentity() bool {
+	return !(k.UserAssignedIDEnabled() && strings.Contains(k.UserAssignedID, "/"))
```
Comment (xuto2) [inline pkg/api/types.go:1974]: why do we need this strings.Contains(k.UserAssignedID, "/") check here but not inside UserAssignedIDEnabled()?

File: pkg/engine/armresources.go [hunk 2]
```diff
@@ -27,24 +27,30 @@ func GenerateARMResources(cs *api.ContainerService) []interface{} {
...
 		useManagedIdentity = kubernetesConfig.UseManagedIdentity
-		userAssignedIDEnabled = useManagedIdentity && kubernetesConfig.UserAssignedID != ""
+		userAssignedIDEnabled = kubernetesConfig.UserAssignedIDEnabled()
+		createNewUserAssignedIdentity = kubernetesConfig.ShouldCreateNewUserAssignedIdentity()
 	}
 
 	isHostedMaster := cs.Properties.IsHostedMasterProfile()
 	if userAssignedIDEnabled {
-		userAssignedID := createUserAssignedIdentities()
+		if createNewUserAssignedIdentity {
```
Comment (xuto2) [inline pkg/engine/armresources.go:41]: I'm curious, in which case do we need to create a new identity when there is already one asssigned by user?

File: pkg/api/types.go (hunk 1 above)
Comment (xuto2) [reply pkg/api/types.go:1974]: ok. so this means it already exists in arm

File: pkg/engine/armvariables.go [hunk 3]
```diff
@@ -67,15 +67,22 @@ func getK8sMasterVars(cs *api.ContainerService) (map[string]interface{}, error)
...
 		enableEncryptionWithExternalKms = to.Bool(kubernetesConfig.EnableEncryptionWithExternalKms)
 		useInstanceMetadata = kubernetesConfig.UseInstanceMetadata
 		excludeMasterFromStandardLB = to.Bool(kubernetesConfig.ExcludeMasterFromStandardLB)
 		maxLoadBalancerCount = kubernetesConfig.MaximumLoadBalancerRuleCount
 		provisionJumpbox = kubernetesConfig.PrivateJumpboxProvision()
+
+		if kubernetesConfig.ShouldCreateNewUserAssignedIdentity() {
+			userAssignedIDReference = "[resourceId('Microsoft.ManagedIdentity/userAssignedIdentities/', variables('userAssignedID'))]"
+		} else {
+			userAssignedIDReference = "[variables('userAssignedID')]"
```
Comment (xuto2) [inline pkg/engine/armvariables.go:84]: The userAssignedID here is a full ARM resource ID, but not something that starts with "Microsoft.ManagedIdentity/userAssignedIdentities", I assume arm can handle both cases?

**Pull request merged**